### ISSUE TITLE
using float and comparison operator

### ISSUE THREAD

**Issue by RogMoe** [RobTillaart/Max44009#7]
I'm automating my roller shutters using time, temp/Rh and lux readings from various sensors. I use your RunninMedian library to smoothen the value of a MAX44009 lux sensor. When I try to compare the lux value to -for example- hours from a ds3231, the output goes bananas. I supsect it's got somethin to do with the float. serialPrint shows an ok value for mediaLux.
Is it possible to translate the float to a unsigned long and would that solve my problem?

`if (medianLux <= 20) {do something; }`

**Comment by RobTillaart** [comment]
Hi Rogmoe,

I do not understand the problem in detail. 
Can you post your sketch, preferably stripped to the essence what is going wrong.

Comparing different datatypes is normally not a problem unless your are out of range for one of the data types.

From your text I understand you are comparing two different physical units.  Light and Time.
In Einsteins universe they might be comparable however not in a trivial way.

Can you explain what you try to accomplish with the comparison ?

**Comment by RogMoe** [comment]
if (now.hour() == 9 && now.minute() == 00 && now.second() == 0) {   // alle rolluiken open om 09:00
    shutters1.setLevel(0);
    shutters2.setLevel(0);
    shutters3.setLevel(0);
    shutters4.setLevel(0);
  }

  if (medianLux < 20) {                                             // als het donker wordt
    shutters1.setLevel(100);
    shutters2.setLevel(100);
    shutters3.setLevel(100);
    shutters4.setLevel(100);
  }

Time comparison works just fine. It moves the shutters down. Lux doesn't. 0=up, 100 fully closed.

See code (work in progress...)
[rolluiken-4st-C.zip](https://github.com/RobTillaart/Max44009/files/7490465/rolluiken-4st-C.zip)

**Comment by RobTillaart** [comment]
OK,

It could be a hysteresis problem. 
The solution is that you should close the shutters at a different light level than the levels you use to open it. 
If you use the same levels and the light is fluctuating around that point it will constantly open /close like madness.

try this:
```cpp
if (medianLux < 20)  closeShutters();
if (medianLux > 30) openShutters();
```

(Your from NL or  BE?)

**Comment by RobTillaart** [comment]
Wow quite some lines of  code!

just a small remark with respect to timing. 
Consider to leave out the seconds just in case the loop() or one of the functions takes a bit longer than expected.
It would also make the comment valid ;)

```cpp
  if (now.hour() == 9 && now.minute() == 00 ) {   // alle rolluiken open om 09:00
```

**Comment by RogMoe** [comment]
NL ;-)

   if (medianLux < 20)  shutters1.setLevel(100);
   if (medianLux > 30) shutters1.setLevel(0);

...makes shutter1 go up and down constantly. Current lux reading is 23, so the shutter should stay in its current state, right? 
Serials shows:
shutters1 going up
shutters2 going up.
shutters3 going up.
shutters4 going up.
shutters1 halted
shutters1 at 0%
shutters2 halted.
shutters2 at 0%
shutters3 halted.
shutters3 at 0%
shutters4 halted.
shutters4 at 0%              (end of initialization)
shutters1 going down     (command issued by lux statement. other shutters stay put.)
shutters1 at 1%
shutters1 at 2%
shutters1 at 3%
shutters1 at 4%
shutters1 at 5%
shutters1 at 6%
...
shutters1 at 98%
shutters1 at 99%
shutters1 halted
shutters1 at 100%
shutters1 going up
shutters1 at 99%
shutters1 at 98%
...
shutters1 at 80%
shutters1 at 79%
shutters1 at 78%
etc etc and keeps going up and down.

If i disable the lux statement line, manual buttons work just fine. I read that using floats in a comparison is bad practise due to it's rounding error. 

(stripped the output a bit to essence)

**Comment by RobTillaart** [comment]
Another cause to consider is that at 9:00 you want to open the shutters (because of time)
but light outside is not enough so that would close it again.

The solution is to combine the conditions

```cpp
  DateTime now = RTC.now();
  if ( ((9 <= now.hour()) &&  (now.hour() < 17)) || (medianLux > 25)) 
 {
  // open the shutters
 }
  if ( (now.hour() <= 9) ||  (now.hour() > 17) || (medianLux < 20) ) 
 {
  // close the shutters
 }
```

**Comment by RobTillaart** [comment]
> I read that using floats in a comparison is bad practise due to it's rounding error.

Yes floats should be compared with inequality tests   <=  or  < or > or >=  

to test two floats for equality should be done with care.
```cpp
float f = 3.14159265;
float g = 22.0/7.0;    // 3.142856...

if ( abs(f - g) < 0.01) Serial.println("equal!!!");
```

and this is only true if  the floats f and g are large compared to the **epsilon** 0.01
```cpp
float f = 0.00001;
float g = 0.00002;

if ( abs(f - g) < 0.01) Serial.println("equal!!!");  // incorrect/
```

**Comment by RogMoe** [comment]
It's a work in progress. It's not all my own code. Needs a fair bit of cleaning, more millis and less delays (if any left at all) etc. Fun part is commenting and organizing. Save that for last. Add a watchdog too, connect resetlines.

**Comment by RobTillaart** [comment]
You should remove all delay's if possible to get "real time" = responsive behavior.

**Comment by RogMoe** [comment]
I was aware that some conditional statements are conflicting. I'm thinking about using a statemachine rather then if/else statements. I was first testing the variables to see if everything works as expected. That's how I stumbled upon the lux-problem.

**Comment by RobTillaart** [comment]
State machine are very powerful especially for these kind of projects. 
Each shutter has 4 states { OPEN, OPENING, CLOSED, CLOSING };   // you might add INIT if you don't know its position..

**Comment by RogMoe** [comment]
I changed the code accordingly, using the and/or operators. It did not change the behaviour. Removed the lux-part just for testing and it turnes out that it stille moves up and down constantly. I guess you were right about the hysteresis. I'm calling the setLevels too many times since it's in the loop. I'll have to sort that first, probably using some flags that get cleared after the movement finishes. Let's fix that first! Back in a sec :-)
